src/node.js

```diff
@@ -22,16 +23,30 @@ export let internalStateKeys = {
   afterSearch: null,
 }
 
-export let initNode = (node, path, extend, types) => {
+export let autoKey = x => F.compactJoin('-', [x.field, x.type])
+
+export let initNode = (
+  node,
+  parentPath,
+  extend,
+  types,
+  dedupe = _.identity
+) => {
   runTypeFunction(types, 'init', node, extend)
-  extend(node, {
-    ..._.omit(_.keys(node), defaults),
-    ..._.omit(_.keys(node), getTypeProp(types, 'defaults', node)),
-    path,
-  })
+  if (node.key) dedupe(node.key) // add node.key to the dedupe cache
+  let key =
+    node.key ||
+    dedupe(runTypeFunctionOrDefault(autoKey, types, 'autoKey', node, extend))
```

Comment:
we probably could have just reused `getTypeProp` instead of creating `runTypeFunctionOrDefault`

Reply:
Doing this with `getTypeProp` basically comes out to `(getTypeProp(types, 'autokey', node) || autoKey)(node)` -- which is just the body of `runTypeFunction` except with `autoKey` as the default function instead of `_.stubTrue`.

It was such a trivial change to generalize it to `runTypeFunctionOrDefault` that even with only one use, I still think it ends up cleaner than the `getTypeProp` alternative